### PR TITLE
Exclude the metrics APIs from resources discovery

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR refactors the Kubernetes resources discovery to exclude the `*.metrics.k8s.io` API groups and thus avoiding querying the cluster metrics and failing to sync when the metrics providers are unavailable.

Tested on EKS with metrics server and by @texasbobs with Prometheus adapter.

Fix: #2605 
Fix: #2120
Fix: #2192

### CODE REVIEW

**Comment by 2opremio** [comment]
Thanks @stefanprodan !

It looks good, however I am not sure whether the specific functions you are calling are cached (I vaguely recall that the memory-cache implementation of the discovery interface only applied caching to certain functions), can you check?

**Comment by stefanprodan** [comment]
It uses the same client as before https://github.com/kubernetes/client-go/blob/master/discovery/discovery_client.go 

When enabling the go client logging the caller is:
```
kubernetes_caller=k8s.io/client-go@v11.0.0+incompatible/discovery/cached/memory/memcache.go
```

File: pkg/cluster/kubernetes/sync.go [hunk 1]
```diff
@@ -239,6 +240,9 @@ func (c *Cluster) getAllowedResourcesBySelector(selector string) (map[string]*ku
 	}
 
 	for _, resource := range resources {
+		if resource == nil {
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync.go]: Why is this needed?

File: pkg/cluster/kubernetes/sync.go [hunk 2]
```diff
@@ -208,21 +207,23 @@ func (c *Cluster) getAllowedResourcesBySelector(selector string) (map[string]*ku
...
+	sgs, err := c.client.discoveryClient.ServerGroups()
+	if sgs == nil {
+		return nil, err
+	}
+
+	resources := []*meta_v1.APIResourceList{}
+	for i := range sgs.Groups {
+		gv := sgs.Groups[i].PreferredVersion.GroupVersion
+		// exclude the *.metrics.k8s.io resources to avoid querying the cluster metrics
+		if !strings.Contains(sgs.Groups[i].Name, "metrics.k8s.io") {
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync.go]: `Contains` seems a little scattershot, can it be more specific? For example, will the name _start_ with metrics.k8s.io; or, if it's a "subdomain", will it _end_ with `.metrics.k8s.io`.

File: pkg/cluster/kubernetes/sync.go (hunk 1 above)
Comment (stefanprodan) [reply pkg/cluster/kubernetes/sync.go]: To avoid panics, a resource can be nil. I've hit this while testing it, I can remove it if you think it's not worth it.

Comment (squaremo) [reply pkg/cluster/kubernetes/sync.go]: The client can return nil entries? Eugh. OK, that's sensible then -- maybe signpost it with a brief comment.

File: pkg/cluster/kubernetes/sync.go (hunk 2 above)
Comment (stefanprodan) [reply pkg/cluster/kubernetes/sync.go]: Doing two comparisons (starts with and ends with) is faster than contains? The `metics.k8s.io` domain is reserved to Kubernetes aggregator API, I don't see how is a scattershot.

Comment (squaremo) [reply pkg/cluster/kubernetes/sync.go]: > Doing two comparisons (starts with and ends with) is faster than contains?

I'm not suggesting doing two comparisons; I'm suggesting being more precise with the one comparison.

> The `metrics.k8s.io` domain is reserved to Kubernetes aggregator API

If that's the exact name, then test for equality. If the items you want to avoid will be `<something>.metrics.k8s.io`, test that they end with `.metrics.k8s.io`. Otherwise it could exclude some future group "whatever-metrics.k8s.io" inadvertently. Sure, it is not that likely such a group would be introduced, but: better to succeed precisely than to fail accidentally.

File: pkg/cluster/kubernetes/sync.go [hunk 3]
```diff
@@ -208,21 +207,23 @@ func (c *Cluster) getAllowedResourcesBySelector(selector string) (map[string]*ku
...
+	sgs, err := c.client.discoveryClient.ServerGroups()
+	if sgs == nil {
+		return nil, err
+	}
+
+	resources := []*meta_v1.APIResourceList{}
+	for i := range sgs.Groups {
+		gv := sgs.Groups[i].PreferredVersion.GroupVersion
+		// exclude the *.metrics.k8s.io resources to avoid querying the cluster metrics
+		if sgs.Groups[i].Name != "metrics.k8s.io" && !strings.HasSuffix(sgs.Groups[i].Name, ".metrics.k8s.io") {
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync.go]: Oh, it can be _either_ of exactly metrics.k8s.io or a subdomain. Eesh. Well this is still better, I reckon -- thanks for following my suggestion up.

**Comment by stefanprodan** [comment]
@squaremo anything else blocking this? I've rebase with master.

File: pkg/cluster/kubernetes/sync.go [hunk 4]
```diff
@@ -208,21 +207,23 @@ func (c *Cluster) getAllowedResourcesBySelector(selector string) (map[string]*ku
...
+			if r, err := c.client.discoveryClient.ServerResourcesForGroupVersion(gv); err == nil {
+				resources = append(resources, r)
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync.go]: Could you just not append `r` if it's `nil`, so you don't need to guard that later?

Comment (stefanprodan) [reply pkg/cluster/kubernetes/sync.go]: done
